### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by SimonSchick**
This is a total conversion to TS (minus jojen stuff), meaning a major release.

This should be completely compatible with the previous version, some things have been convert to classes but overall the data structures are the same.

### CODE REVIEW

File: .npmignore [hunk 1]
```diff
@@ -1,5 +1,9 @@
...
-.eslintrc
-.travis.yml
 *.sublime-project
 *.sublime-workspace
+
+go
+src
+tslint.json
+.gitignore
+/.vscode
```
Comment (connor4312) [inline .npmignore]: Please prefix all ignores with a slash when applicable to prevent surprising behaviour 😄

File: go/bcommon/roles.go [hunk 2]
```diff
@@ -132,153 +132,6 @@ var (
...
-			UserDelete,
-			UserDetails,
-			UserGetDiscordInvite + ":self",
-			UserLog,
-			UserNotification + ":self",
-			UserSeen + ":self",
-			UserTwofactor,
-			UserUpdate,
-		},
-	}
```
Comment (connor4312) [inline go/bcommon/roles.go]: Something isn't working right, porting the Js side to TypeScript should cause no changes in the generated Go files.

File: src/emoticons.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        if (count >= 250) { return 500 - count; }
+        if (count >= 150) { return 250 - count; }
+        if (count >= 100) { return 150 - count; }
+        if (count >= 60) { return 100 - count; }
+        if (count >= 30) { return 60 - count; }
+        if (count >= 15) { return 30 - count; }
+        if (count >= 5) { return 15 - count; }
+        return 5 - count;
+    }
+}
```
Comment (connor4312) [inline src/emoticons.ts]: I realize you were just porting over, but these could probably be simplified by making a single `{ minSubs: number, unlockedEmoticons: number }[]` 😛

File: src/roles/RoleManager.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    };
+
+    /**
+     * Returns whether one `role` can administrate an `other`. This will be
+     * true if the first role has a level greater than 25, and has a level
+     * at least ten greater than the other.
+     *
+     * @param  {String} role
+     * @param  {String} other
+     * @return {Boolean}
```
Comment (connor4312) [inline src/roles/RoleManager.ts]: jsdoc types can be 🔥  unless we want to keep them for possible esdoc generation

**Comment by connor4312** [comment]
Assuming you just did a 1:1 port for the most part. Difference in the generated Go is my main concern.

File: src/roles/RoleManager.ts (hunk 4 above)
Comment (SimonSchick) [inline src/roles/RoleManager.ts]: I thought I caught em all 😛

**Comment by SimonSchick** [comment]
Fixed the problem with the go file, unfortunately the order changed, I can fix that if it's still a concern.

**Comment by SimonSchick** [comment]
Fixed ordering, `go` files are untouched 😄

**Comment by miyu** [comment]
Have you npm linked with BE to run tests?

**Comment by SimonSchick** [comment]
I have,
there is 3 incompatibilities in BE:

The diff for BE to work would look like this:

```diff
diff --git a/api/app/oauth/OAuthAuthorizationHandler.js b/api/app/oauth/OAuthAuthorizationHandler.js
index a884d13..ef62db7 100644
--- a/api/app/oauth/OAuthAuthorizationHandler.js
+++ b/api/app/oauth/OAuthAuthorizationHandler.js
@@ -41,7 +41,7 @@ class OAuthAuthorizationHandler {
             _(user.groups)
             .map('name')
             .concat(CHANNEL_ROLES)
-            .map(roles.get)
+            .map(name => roles.get(name))
             .value()
         );

diff --git a/api/app/preferences/preferences.js b/api/app/preferences/preferences.js
index 369f885..64e33f0 100644
--- a/api/app/preferences/preferences.js
+++ b/api/app/preferences/preferences.js
@@ -3,7 +3,7 @@
 const Bluebird = require('bluebird');
 const Result = require('../../arch/result');

-const listing = require('@mcph/beam-common/preferences');
+const listing = require('@mcph/beam-common').preferences;
 const db = require('../../db/sql');
 const _ = require('lodash');

diff --git a/api/arch/PermissionChecker.js b/api/arch/PermissionChecker.js
index 046aeeb..366cfca 100644
--- a/api/arch/PermissionChecker.js
+++ b/api/arch/PermissionChecker.js
@@ -111,7 +111,7 @@ class PermissionChecker {
      * @returns {String[]}
      */
     _getPermissionsUnionOfGroups (groups) {
-        return roles.getPermissions(_(groups).map('name').map(roles.get).value());
+        return roles.getPermissions(_(groups).map('name').map(name => roles.get(name)).value());
     }

     /**
```

1) `common` via `const listing = require('@mcph/beam-common/preferences');` which is generally a bad idea.
2) Passing `roles.get` as a anonymous function causing it to lose it's `this` context.
3) See above.

As this would be a major release anything using common would not be affected by a publish.

Since this is a `major` release I'm considering to not make `common` 100% compatible as the above problems are usually caused by bad practise.

**Comment by SimonSchick** [comment]
Opinions @connor4312 @miyu ?

**Comment by miyu** [comment]
+1

**Pull request merged**